pkg/controller/direct/notebooks/instance_controller.go

```diff
@@ -0,0 +1,269 @@
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewInstanceIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get notebooks GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &InstanceAdapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *modelInstance) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type InstanceAdapter struct {
+	id        *krm.InstanceIdentity
+	gcpClient *gcp.NotebookClient
+	desired   *krm.NotebookInstance
+	actual    *notebookspb.Instance
+}
+
+var _ directbase.Adapter = &InstanceAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *InstanceAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting Instance", "name", a.id)
+
+	req := &notebookspb.GetInstanceRequest{Name: a.id.String()}
+	instancepb, err := a.gcpClient.GetInstance(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting Instance %q: %w", a.id, err)
+	}
+
+	a.actual = instancepb
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *InstanceAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating Instance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := NotebookInstanceSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &notebookspb.CreateInstanceRequest{
+		Parent:     a.id.Parent().String(),
+		InstanceId: a.id.ID(),
+		Instance:   resource,
+	}
+	op, err := a.gcpClient.CreateInstance(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Instance %s: %w", a.id, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("instance %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created Instance", "name", a.id)
+
+	status := &krm.NotebookInstanceStatus{}
+	status.ObservedState = NotebookInstanceObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *InstanceAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
```

Comment:
We are not testing the update? Can we add an update.yaml to change something simple, like "labels"?